dask/dataframe/tests/test_groupby.py

```diff
@@ -26,8 +107,8 @@ def func(df):
 
     assert eq(d.groupby('a').apply(func), full.groupby('a').apply(func))
 
-    assert sorted(d.groupby('a').apply(func).dask) == \
-           sorted(d.groupby('a').apply(func).dask)
+    # assert sorted(d.groupby('a').apply(func).dask) == \
```

Comment:
OK, mostly done except for one point. I've changed ``groupby.apply`` to actually use ``shuffle`` to get correct result. However, the change makes the above test fail. 

The failure is because ``shuffle (set_partition)`` has a ``barrier`` logic which always use new token. I don't know the background, but can it be the same key if input is the same?

- https://github.com/dask/dask/blob/d9b8a146f17455780ed553df344396fbca0bdbd0/dask/dataframe/shuffle.py#L116

## Previous logic

Previous logic doesn't actually use shuffled data, so the test has been passed (but lead to incorrect results).

```
            df = shuffle(self.df, self.index, **self.kwargs)
            return map_partitions(_groupby_apply,
                                  columns or self.df.columns,
                                  self.df, self.index, func)
```